examples/plot_anomaly_comparison.py

```diff
@@ -0,0 +1,123 @@
+but that in practice they need to be adjusted. In the absence of labelled
+data, the problem is completely unsupervised so model selection can be
+a challenge.
+"""
+print(__doc__)  # noqa
+
+import time
+
+import numpy as np
+from scipy import stats
+import matplotlib
+import matplotlib.pyplot as plt
+
+from sklearn import svm
+from sklearn.datasets import make_moons, make_blobs
+from sklearn.covariance import EllipticEnvelope
+from sklearn.ensemble import IsolationForest
+from sklearn.neighbors import LocalOutlierFactor
+
+matplotlib.rcParams['contour.negative_linestyle'] = 'solid'
+
+# Example settings
+n_samples = 300
+outliers_fraction = 0.15
+n_outliers = int(outliers_fraction * n_samples)
+n_inliers = n_samples - n_outliers
+
+# define outlier/anomaly detection methods to be compared
+anomaly_algorithms = [
+    ("Robust covariance", EllipticEnvelope(contamination=outliers_fraction)),
+    ("One-Class SVM", svm.OneClassSVM(nu=0.2, kernel="rbf", gamma=0.1)),
+    ("Isolation Forest", IsolationForest(contamination=outliers_fraction,
+                                         random_state=42)),
+    ("Local Outlier Factor", LocalOutlierFactor(
+        n_neighbors=35, contamination=outliers_fraction))]
+
+# Define datasets
+blobs_params = dict(random_state=0, n_samples=n_inliers, n_features=2)
+datasets = [
+    make_blobs(centers=[[0, 0], [0, 0]], cluster_std=0.5,
+               **blobs_params)[0],
+    make_blobs(centers=[[2, 2], [-2, -2]], cluster_std=[1.5, .3],
+               **blobs_params)[0],
+    4. * (make_moons(n_samples=n_samples, noise=.05)[0] -
+          np.array([0.5, 0.25])),
+    14. * (np.random.RandomState(42).rand(n_samples, 2) - 0.5)]
+
+# Compare given classifiers under given settings
+xx, yy = np.meshgrid(np.linspace(-7, 7, 150),
+                     np.linspace(-7, 7, 150))
+
+plt.figure(figsize=(len(anomaly_algorithms) * 2 + 3, 12.5))
+plt.subplots_adjust(left=.02, right=.98, bottom=.001, top=.96, wspace=.05,
+                    hspace=.01)
+
+plot_num = 1
+
+for i_dataset, X in enumerate(datasets):
+    # Add outliers
+    rng = np.random.RandomState(42)
+    X = np.concatenate([X, rng.uniform(low=-6, high=6,
+                       size=(n_outliers, 2))], axis=0)
+
+    for name, algorithm in anomaly_algorithms:
+        t0 = time.time()
+        algorithm.fit(X)
+        t1 = time.time()
+        plt.subplot(len(datasets), len(anomaly_algorithms), plot_num)
+        if i_dataset == 0:
+            plt.title(name, size=18)
+
+        # fit the data and tag outliers
+        if name == "Local Outlier Factor":
+            anomaly_scores = algorithm.negative_outlier_factor_
+        else:
+            anomaly_scores = algorithm.decision_function(X).ravel()
+
+        threshold = stats.scoreatpercentile(anomaly_scores,
+                                            100 * outliers_fraction)
+        y_pred = (anomaly_scores >= threshold).astype(int)
```

Comment:
I think it's a bit weird to reimplement predict, but whatever... lgtm

Reply:
Hm I suppose @agramfort and I are a bit biased about how the OCSVM should be used :)
This is the plot we have if we use predict only and nu=outliers_fraction (instead of nu=0.2). Its not good for the OCSVM but I guess the users are going to use the algorithms this way so if there is a consensus I’m OK with this.
![figure_1](https://user-images.githubusercontent.com/15966638/32067390-57386ae8-ba83-11e7-93f8-113c6bda292e.png)

Reply:
Can you elaborate? If you're saying OCSVM should not be used the way it is implemented, that seems pretty bad?

Reply:
In particular if the example shows a usage pattern that is not explained somewhere else, that seems pretty confusing.

Reply:
The `nu` parameter of the OCSVM does not allow to control the proportion of outliers as the `contamination` parameter does for the other algorithms. The `contamination` parameter of the other algorithms allows to define a threshold such that we exactly predict a proportion of outliers equal to `contamination`. The OCSVM automatically finds an offset/threshold which is such that the proportion of outliers is less than `nu` (but not necessarily equal to nu depending on the approximation/number of samples).
I prefer to choose a value for `nu` a bit higher than the proportion of outliers that I want to find and then threshold the decision function to obtain the desired proportion of outliers (as it is done for the other algorithms). Choosing a value for `nu` a bit higher than the proportion of outliers gives a bit more support vectors to approximate the decision function in the neighborhood of the decision boundary. But this extends the original implementation of the OCSVM and does not use the learned offset.